cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -575,16 +588,27 @@ public String getOutputFormatClassName() {
       throw new UnsupportedOperationException(
         "Output is not supported for external partitioned file set '" + spec.getName() + "'");
     }
+
+    // copy the output properties of the embedded file set to the output arguments
+    Map<String, String> outputArgs = Maps.newHashMap(files.getOutputFormatConfiguration());
+
     // we set the file set's output path in the definition's getDataset(), so there is no need to configure it again.
-    // here we just want to validate that an output partition key was specified in the arguments.
+    // here we just want to validate that an output partition key or dynamic partitioner was specified in the arguments.
     PartitionKey outputKey = PartitionedFileSetArguments.getOutputPartitionKey(runtimeArguments, getPartitioning());
-    if (outputKey == null) {
-      throw new DataSetException("Partition key must be given for the new output partition as a runtime argument.");
+    if (outputKey != null) {
+      PartitionedFileSetArguments.setOutputPartitionKey(outputArgs, outputKey);
+    } else {
+      String dynamicPartitionerClassName = PartitionedFileSetArguments.getDynamicPartitioner(runtimeArguments);
+      if (dynamicPartitionerClassName == null) {
+        throw new DataSetException(
+          "Either a Partition key or a DynamicPartitioner class must be given as a runtime argument.");
+      }
+      // also delegate to the files.getOutputFormatClassName()? (currently defaulting to TextOutputFormat)
+      PartitionedFileSetArguments.setDynamicPartitioner(outputArgs, dynamicPartitionerClassName);
+      outputArgs.put("output.format.class.name", files.getOutputFormatClassName());
```

Comment:
The constants were defined in app-fabric (where the DynamicPartitioningOutputFormat is).
I moved the constants to Constants.Dataset.Partitioned for usage from both places.